### PR TITLE
feat: manage_group_duty_snap周期任务出现死锁 --story=118055769

### PR DESCRIPTION
**Pull request by lthon-sha**

### CODE REVIEW

**Comment by lthon-sha** [comment]
------------------------
LATEST DETECTED DEADLOCK
------------------------
2024-06-18 20:22:00 0x7f3171537700
*** (1) TRANSACTION:
TRANSACTION 494835094, ACTIVE 0 sec starting index read
mysql tables in use 4, locked 3
LOCK WAIT 4 lock struct(s), heap size 1136, 3 row lock(s)
MySQL thread id 97912116, OS thread handle 139852910384896, query id 11374058063 172.17.232.74 root Sending data
SELECT `bkmonitor_dutyplan`.`id`, `bkmonitor_dutyplan`.`user_group_id`, `bkmonitor_dutyplan`.`duty_rule_id`, `bkmonitor_dutyplan`.`duty_arrange_id`, `bkmonitor_dutyplan`.`order`, `bkmonitor_dutyplan`.`user_index`, `bkmonitor_dutyplan`.`is_active`, `bkmonitor_dutyplan`.`is_effective`, `bkmonitor_dutyplan`.`start_time`, `bkmonitor_dutyplan`.`finished_time`, `bkmonitor_dutyplan`.`users`, `bkmonitor_dutyplan`.`work_times`, `bkmonitor_dutyplan`.`timezone`, `bkmonitor_dutyplan`.`begin_time`, `bkmonitor_dutyplan`.`end_time`, `bkmonitor_dutyplan`.`last_send_time`, `bkmonitor_dutyplan`.`duty_time` FROM `bkmonitor_dutyplan` WHERE (`bkmonitor_dutyplan`.`duty_rule_id` IN (SELECT U0.`id` FROM `duty_rule` U0 WHERE (NOT U0.`is_deleted` AND U0.`bk_biz_id` = 2 AND NOT U0.`enabled`)) AND `bkmonitor_dutyplan`.`is_effective` = 1 AND `bkmonitor_dutyplan`.`user_group_id` = 156) FOR UPDATE
*** (1) WAITING FOR THIS LOCK TO BE GRANTED:
RECORD LOCKS space id 2595 page no 44 n bits 1000 index bkmonitor_dutyplan_is_effective_08a1f769 of table `bk_monitor`.`bkmonitor_dutyplan` trx id 494835094 lock_mode X locks rec but not gap waiting
Record lock, heap no 97 PHYSICAL RECORD: n_fields 2; compact format; info bits 0
 0: len 4; hex 80000001; asc     ;;
 1: len 8; hex 800000000000001f; asc         ;;

*** (2) TRANSACTION:
TRANSACTION 494835093, ACTIVE 0 sec fetching rows
mysql tables in use 4, locked 3
15 lock struct(s), heap size 1136, 73 row lock(s)
MySQL thread id 97912115, OS thread handle 139850331420416, query id 11374058062 172.17.232.74 root Sending data
SELECT `bkmonitor_dutyplan`.`id`, `bkmonitor_dutyplan`.`user_group_id`, `bkmonitor_dutyplan`.`duty_rule_id`, `bkmonitor_dutyplan`.`duty_arrange_id`, `bkmonitor_dutyplan`.`order`, `bkmonitor_dutyplan`.`user_index`, `bkmonitor_dutyplan`.`is_active`, `bkmonitor_dutyplan`.`is_effective`, `bkmonitor_dutyplan`.`start_time`, `bkmonitor_dutyplan`.`finished_time`, `bkmonitor_dutyplan`.`users`, `bkmonitor_dutyplan`.`work_times`, `bkmonitor_dutyplan`.`timezone`, `bkmonitor_dutyplan`.`begin_time`, `bkmonitor_dutyplan`.`end_time`, `bkmonitor_dutyplan`.`last_send_time`, `bkmonitor_dutyplan`.`duty_time` FROM `bkmonitor_dutyplan` WHERE (`bkmonitor_dutyplan`.`duty_rule_id` IN (SELECT U0.`id` FROM `duty_rule` U0 WHERE (NOT U0.`is_deleted` AND U0.`bk_biz_id` = 2 AND NOT U0.`enabled`)) AND `bkmonitor_dutyplan`.`is_effective` = 1 AND `bkmonitor_dutyplan`.`user_group_id` = 153) FOR UPDATE
*** (2) HOLDS THE LOCK(S):
RECORD LOCKS space id 2595 page no 44 n bits 1000 index bkmonitor_dutyplan_is_effective_08a1f769 of table `bk_monitor`.`bkmonitor_dutyplan` trx id 494835093 lock_mode X locks rec but not gap
Record lock, heap no 97 PHYSICAL RECORD: n_fields 2; compact format; info bits 0
 0: len 4; hex 80000001; asc     ;;
 1: len 8; hex 800000000000001f; asc         ;;

Record lock, heap no 98 PHYSICAL RECORD: n_fields 2; compact format; info bits 0
 0: len 4; hex 80000001; asc     ;;
 1: len 8; hex 8000000000000020; asc         ;;

Record lock, heap no 99 PHYSICAL RECORD: n_fields 2; compact format; info bits 0
 0: len 4; hex 80000001; asc     ;;
 1: len 8; hex 8000000000000021; asc        !;;

Record lock, heap no 100 PHYSICAL RECORD: n_fields 2; compact format; info bits 0
 0: len 4; hex 80000001; asc     ;;
 1: len 8; hex 8000000000000022; asc        ";;

Record lock, heap no 101 PHYSICAL RECORD: n_fields 2; compact format; info bits 0
 0: len 4; hex 80000001; asc     ;;
 1: len 8; hex 8000000000000023; asc        #;;

Record lock, heap no 102 PHYSICAL RECORD: n_fields 2; compact format; info bits 0
 0: len 4; hex 80000001; asc     ;;
 1: len 8; hex 8000000000000024; asc        $;;

Record lock, heap no 103 PHYSICAL RECORD: n_fields 2; compact format; info bits 0
 0: len 4; hex 80000001; asc     ;;
 1: len 8; hex 8000000000000025; asc        %;;

Record lock, heap no 104 PHYSICAL RECORD: n_fields 2; compact format; info bits 0
 0: len 4; hex 80000001; asc     ;;
 1: len 8; hex 8000000000000026; asc        &;;

Record lock, heap no 123 PHYSICAL RECORD: n_fields 2; compact format; info bits 0
 0: len 4; hex 80000001; asc     ;;
 1: len 8; hex 800000000000003e; asc        >;;

Record lock, heap no 124 PHYSICAL RECORD: n_fields 2; compact format; info bits 0
 0: len 4; hex 80000001; asc     ;;
 1: len 8; hex 800000000000003f; asc        ?;;

Record lock, heap no 125 PHYSICAL RECORD: n_fields 2; compact format; info bits 0
 0: len 4; hex 80000001; asc     ;;
 1: len 8; hex 8000000000000040; asc        @;;

Record lock, heap no 126 PHYSICAL RECORD: n_fields 2; compact format; info bits 0
 0: len 4; hex 80000001; asc     ;;
 1: len 8; hex 8000000000000041; asc        A;;

Record lock, heap no 127 PHYSICAL RECORD: n_fields 2; compact format; info bits 0
 0: len 4; hex 80000001; asc     ;;
 1: len 8; hex 8000000000000042; asc        B;;

Record lock, heap no 128 PHYSICAL RECORD: n_fields 2; compact format; info bits 0
 0: len 4; hex 80000001; asc     ;;
 1: len 8; hex 8000000000000043; asc        C;;

Record lock, heap no 129 PHYSICAL RECORD: n_fields 2; compact format; info bits 0
 0: len 4; hex 80000001; asc     ;;
 1: len 8; hex 8000000000000044; asc        D;;

Record lock, heap no 130 PHYSICAL RECORD: n_fields 2; compact format; info bits 0
 0: len 4; hex 80000001; asc     ;;
 1: len 8; hex 8000000000000045; asc        E;;

Record lock, heap no 131 PHYSICAL RECORD: n_fields 2; compact format; info bits 0
 0: len 4; hex 80000001; asc     ;;
 1: len 8; hex 8000000000000046; asc        F;;

Record lock, heap no 132 PHYSICAL RECORD: n_fields 2; compact format; info bits 0
 0: len 4; hex 80000001; asc     ;;
 1: len 8; hex 8000000000000047; asc        G;;

Record lock, heap no 143 PHYSICAL RECORD: n_fields 2; compact format; info bits 0
 0: len 4; hex 80000001; asc     ;;
 1: len 8; hex 8000000000000052; asc        R;;

Record lock, heap no 325 PHYSICAL RECORD: n_fields 2; compact format; info bits 0
 0: len 4; hex 80000001; asc     ;;
 1: len 8; hex 8000000000000114; asc         ;;

Record lock, heap no 326 PHYSICAL RECORD: n_fields 2; compact format; info bits 0
 0: len 4; hex 80000001; asc     ;;
 1: len 8; hex 8000000000000115; asc         ;;

Record lock, heap no 327 PHYSICAL RECORD: n_fields 2; compact format; info bits 0
 0: len 4; hex 80000001; asc     ;;
 1: len 8; hex 8000000000000116; asc         ;;

Record lock, heap no 328 PHYSICAL RECORD: n_fields 2; compact format; info bits 0
 0: len 4; hex 80000001; asc     ;;
 1: len 8; hex 8000000000000117; asc         ;;

Record lock, heap no 337 PHYSICAL RECORD: n_fields 2; compact format; info bits 0
 0: len 4; hex 80000001; asc     ;;
 1: len 8; hex 8000000000000120; asc         ;;

*** (2) WAITING FOR THIS LOCK TO BE GRANTED:
RECORD LOCKS space id 2595 page no 18 n bits 80 index PRIMARY of table `bk_monitor`.`bkmonitor_dutyplan` trx id 494835093 lock_mode X locks rec but not gap waiting
Record lock, heap no 3 PHYSICAL RECORD: n_fields 19; compact format; info bits 0
 0: len 8; hex 8000000000000120; asc         ;;
 1: len 6; hex 0000000032c9; asc     2 ;;
 2: len 7; hex b400000031149a; asc     1  ;;
 3: len 4; hex 8000009c; asc     ;;
 4: SQL NULL;
 5: len 4; hex 80000000; asc     ;;
 6: len 1; hex 80; asc  ;;
 7: len 30; hex 020100750000070005006e0027000200290004002d000400310007003800; asc    u      n '   )   -   1   8 ; (total 118 bytes);
 8: SQL NULL;
 9: SQL NULL;
 10: len 5; hex 0000000400; asc      ;;
 11: len 4; hex 80000015; asc     ;;
 12: len 19; hex 323032332d31322d31342032333a30303a3539; asc 2023-12-14 23:00:59;;
 13: len 4; hex 80000001; asc     ;;
 14: len 19; hex 323032332d31312d31352030303a30303a3030; asc 2023-11-15 00:00:00;;
 15: len 13; hex 417369612f5368616e67686169; asc Asia/Shanghai;;
 16: len 30; hex 0216004a06004600008c0000d200001801005e0100a40100ea0100300200; asc    J  F           ^        0  ; (total 1611 bytes);
 17: len 4; hex 80000000; asc     ;;
 18: len 4; hex 80000000; asc     ;;

*** WE ROLL BACK TRANSACTION (1)

**Comment by lthon-sha** [comment]
```mysql
mysql> explain SELECT `bkmonitor_dutyplan`.`id`, `bkmonitor_dutyplan`.`user_group_id`, `bkmonitor_dutyplan`.`duty_rule_id`, `bkmonitor_dutyplan`.`duty_arrange_id`, `bkmonitor_dutyplan`.`order`, `bkmonitor_dutyplan`.`user_index`, `bkmonitor_dutyplan`.`is_active`, `bkmonitor_dutyplan`.`is_effective`, `bkmonitor_dutyplan`.`start_time`, `bkmonitor_dutyplan`.`finished_time`, `bkmonitor_dutyplan`.`users`, `bkmonitor_dutyplan`.`work_times`, `bkmonitor_dutyplan`.`timezone`, `bkmonitor_dutyplan`.`begin_time`, `bkmonitor_dutyplan`.`end_time`, `bkmonitor_dutyplan`.`last_send_time`, `bkmonitor_dutyplan`.`duty_time` FROM `bkmonitor_dutyplan` WHERE (`bkmonitor_dutyplan`.`duty_rule_id` IN (SELECT U0.`id` FROM `duty_rule` U0 WHERE (NOT U0.`is_deleted` AND U0.`bk_biz_id` = 2 AND NOT U0.`enabled`)) AND `bkmonitor_dutyplan`.`is_effective` = 1 AND `bkmonitor_dutyplan`.`user_group_id` = 156) \G
*************************** 1. row ***************************
           id: 1
  select_type: SIMPLE
        table: U0
   partitions: NULL
         type: ALL
possible_keys: PRIMARY,duty_rule_bk_biz_id_a2e5c104
          key: NULL
      key_len: NULL
          ref: NULL
         rows: 27
     filtered: 3.70
        Extra: Using where
*************************** 2. row ***************************
           id: 1
  select_type: SIMPLE
        table: bkmonitor_dutyplan
   partitions: NULL
         type: index_merge
possible_keys: bkmonitor_dutyplan_user_group_id_a04a0764,bkmonitor_dutyplan_duty_rule_id_19e8401f,bkmonitor_dutyplan_is_effective_08a1f769
          key: bkmonitor_dutyplan_user_group_id_a04a0764,bkmonitor_dutyplan_is_effective_08a1f769
      key_len: 4,4
          ref: NULL
         rows: 42
     filtered: 6.67
        Extra: Using intersect(bkmonitor_dutyplan_user_group_id_a04a0764,bkmonitor_dutyplan_is_effective_08a1f769); Using where; Using join buffer (Block Nested Loop)
2 rows in set, 1 warning (0.00 sec)

**Comment by lthon-sha** [comment]
### trx 5094 (用户组 156)

#### 等待 is_effective 锁 (1, 31)

```mysql
mysql> select * from bkmonitor_dutyplan where id = 31 \G
*************************** 1. row ***************************
             id: 31
  user_group_id: 153
```

### trx 5093 (用户组 153)
#### 持有 is_effective 锁

key 值都为 1， 主键值如下，**除了 288，其余对应的 user_group_id 均为 153**

- 31 到 38 (1f 到 26)
- 62 到 71 (3e 到 47)
- 82 (52)
- 276 到 279 (114 到 117)
- 288 (120)
#### 等待 pk 锁 288 (g 156)

```mysql
mysql> select * from bkmonitor_dutyplan where id = 288 \G
*************************** 1. row ***************************
             id: 288
  user_group_id: 156

**Comment by lthon-sha** [comment]
由于索引合并的加锁方式，导致上锁顺序不一致，在上锁过程中出现死锁。

File: bkmonitor/bkmonitor/models/strategy.py [hunk 1]
```diff
@@ -808,9 +808,9 @@ class DutyPlan(Model):
     is_active = models.BooleanField("是否生效状态（已废弃）", default=False)
 
     # 是否有效，替换原来的is_active字段，这样可以设置索引
-    is_effective = models.IntegerField("是否有效", default=0, db_index=True)
+    is_effective = models.IntegerField("是否有效", default=0)
     start_time = models.CharField("当前轮班生效开始时间", null=False, max_length=32, default="1970-01-01 00:00:00")
-    finished_time = models.CharField("当前轮班生效结束时间", null=True, max_length=32, db_index=True)
```
Comment (chenguo367) [inline bkmonitor/bkmonitor/models/strategy.py:813]: 这个独立索引需要保留的。 联合索引在单独基于finished_time查询时没法支持到。

**Review by chenguo367 — CHANGES_REQUESTED** [review]

File: bkmonitor/bkmonitor/models/strategy.py (hunk 1 above)
Comment (lthon-sha) [reply bkmonitor/bkmonitor/models/strategy.py:813]: done

**Review by chenguo367 — APPROVED** [review]

**Pull request merged**